### PR TITLE
[JW8-10890] Tizen Browser Support (2016, 2018, 2019)

### PR DESCRIPTION
**Pull request by robwalch**
### This PR will...
Address issues that prevent the player from setting up and playing through playlists on 2016 Webkit based Tizen Samsung TV browsers. Old Mac owners who use Safari 5-8 can rejoice as these fixes apply to there as well.

- Add partial webkit flexbox fallbacks (partial, some elements need more work)
- Disables background loading for playlist progression to work (Ideally the user agent browser flag would not be necessary, but this is a quick fix for the purpose of exposing the need for better feature detection and broader support for disabling background loading)
- Disable autoplay test when `new Blob` thows
- Uses polyfill for `parser.parseFromString('', 'text/html')`

### Other issues not addressed
- CSS opacity animations/transitions should be disabled completely
- Motion thumbnail video elements show up black before and after playback (and trick play should be based on background loading)
- Playback toggles to paused when pressing the fullscreen button
- Testing on 2017 models
- Testing in an application bundle
- Testing ads (vast/ima/fw)
- Testing shelf (overlay works fine)
- hls.js support not included in this PR

#### Findings on 2018-2019 Samsung TVs (6-7 Series)
- Background loading must also be disabled to prevent DOM video from going black
- HLS streams are supported with `video.src = '//url.m3u8'`
- The hls.js provider functions correctly when our player recognizes the user-agent as Chrome

### Are there any Pull Requests open in other repos which need to be merged with this?
No.

#### Addresses Issue(s):
[JW8-10890](https://jwplayer.atlassian.net/browse/JW8-10890) / JW8-10701

### CODE REVIEW

File: src/js/providers/html5.js [hunk 1]
```diff
@@ -476,9 +476,16 @@ function VideoProvider(_playerId, _playerConfig, mediaElement) {
...
+                    currentQuality = i;
+                    break;
                 }
             }
+            if (!currentQuality) {
+                // Choose the highest quality
+                const sorted = levels.map((level, index) => Object.assign({ index, height: 0 }, level))
+                    .sort((a, b) => b.height - a.height);
+                currentQuality = sorted[0].index;
+            }
```
Comment (robwalch) [inline src/js/providers/html5.js]: This block was added for demo purposes on the 2016 device. 2018-19 devices support adaptive hls playback with the hls.js provider (0.12.4 fork).

File: src/js/environment/environment.js [hunk 2]
```diff
@@ -175,7 +180,7 @@ Object.defineProperties(Features, {
         enumerable: true
     },
     backgroundLoading: {
-        get: memoize(() => !(OS.iOS || Browser.safari)),
+        get: memoize(() => !(OS.iOS || Browser.safari || OS.tizen)),
```
Comment (robwalch) [inline src/js/environment/environment.js]: This change is critical for ads and playlist playback. The TVs do not handle buffering media into another video element in the background while one is still playing.

File: src/css/controls/imports/controlbar.less [hunk 3]
```diff
@@ -82,13 +82,23 @@
...
     width: 100%;
     padding: 0 12px;
+
+    div {
+        float: left;
+        line-height: 21px;
+    }
+
+    .jw-spacer ~ div {
+        float: right;
```
Comment (robwalch) [inline src/css/controls/imports/controlbar.less]: The control bar buttons on the right will show up in reverse order in old Webkit browsers missing flex-box support, but that's better than having them jammed up against the buttons on the left.

File: src/js/environment/environment.js (hunk 2 above)
Comment (jnatalzia) [reply src/js/environment/environment.js]: nice

File: src/js/environment/os-version.js [hunk 4]
```diff
@@ -31,6 +31,8 @@ export function osVersion(osEnvironment, agent) {
         version = execResult(/OS ([._\d]+)/.exec(agent), 1);
     } else if (osEnvironment.mac) {
         version = execResult(/Mac OS X (10[._\d]+)/.exec(agent), 1);
+    } else if (osEnvironment.tizen) {
```
Comment (jnatalzia) [inline src/js/environment/os-version.js]: if we wanted to add webos, i saw in video-dev it's `Web0S` (yes a zero)

File: src/js/providers/html5.js [hunk 5]
```diff
@@ -476,9 +476,16 @@ function VideoProvider(_playerId, _playerConfig, mediaElement) {
...
         }
         visualQuality.reason = 'initial choice';
```
Comment (jnatalzia) [inline src/js/providers/html5.js]: is it safe to fall through and hit this block? previously the reason would never get set if we found the level above

File: src/js/utils/browser.js [hunk 6]
```diff
@@ -32,8 +32,13 @@ export function isMSIE() {
     return userAgentMatch(/msie/i);
 }
 
+export function isTizen() {
+    return userAgentMatch(/SMART-TV/);
```
Comment (jnatalzia) [inline src/js/utils/browser.js]: this is different than the approach above where we look for `Tizen`, can i get a quick explainer?

File: src/js/utils/can-autoplay.js [hunk 7]
```diff
@@ -29,6 +29,7 @@ const URL = (global => global.URL && global.URL.createObjectURL ?
     global.URL : global.webkitURL || global.mozURL)(window);
 
 let videoBlob;
+const Blob = window.Blob;
```
Comment (jnatalzia) [inline src/js/utils/can-autoplay.js]: i assume this is what the new lint rule is for? is it worth calling it something diff? are there risks with having a var the same name as a window var?

File: src/js/providers/html5.js (hunk 5 above)
Comment (robwalch) [reply src/js/providers/html5.js]: Yeah. That was a mistake.

File: src/js/utils/browser.js (hunk 6 above)
Comment (robwalch) [reply src/js/utils/browser.js]: "Tizen" is not present in the pre-2017 Smart TV user-agents. `SMART-TV` indicated it is Tizen. The word _Tizen_ is found before the Tizen _version_ in later models.

File: src/js/utils/can-autoplay.js (hunk 7 above)
Comment (robwalch) [reply src/js/utils/can-autoplay.js]: Nothing wrong with shadowing global names. We should always get globals from the global object (`window` or `self`).

File: src/css/controls/imports/controlbar.less [hunk 8]
```diff
@@ -82,13 +82,23 @@
 }
 
 .jw-button-container {
+    display: block;
```
Comment (robwalch) [inline src/css/controls/imports/controlbar.less]: Would be better to use autoprerixer to compile backwards compatible support:

From https://autoprefixer.github.io/
```css
    display: -webkit-box;
    display: flex;
    -webkit-box-orient: horizontal;
    -webkit-box-direction: normal;
            flex-flow: row nowrap;
    -webkit-box-flex: 1;
            flex: 1 1 auto;
    -webkit-box-align: center;
            align-items: center;
    -webkit-box-pack: center;
            justify-content: center;
    width: 100%;
    padding: 0 12px;
```

**Comment by robwalch** [comment]
Test this please

File: src/css/controls/imports/controlbar.less (hunk 8 above)
Comment (jnatalzia) [reply src/css/controls/imports/controlbar.less]: sounds like you'd prefer a manual addition where needed? i tend to agree, autoprefixr could really bloat

File: src/css/controls/imports/settings-menu.less [hunk 9]
```diff
@@ -135,6 +136,10 @@
     .jw-settings-close {
         margin-left: auto;
     }
+
+    div {
```
Comment (jnatalzia) [inline src/css/controls/imports/settings-menu.less]: lots of generic selectors, might bite us in future but seems safe enough given they are localized in these blocks

File: src/js/utils/browser.ts [hunk 10]
```diff
@@ -15,12 +15,12 @@ export function isFlashSupported(): boolean {
     return !!(version && version >= __FLASH_VERSION__);
 }
 
+const isIPadOS13 = () => navigator.platform === 'MacIntel' && navigator.maxTouchPoints > 1;
```
Comment (jnatalzia) [inline src/js/utils/browser.ts]: Out of curiosity what’s the reason these are functions vs set consts

File: src/js/utils/dom.js [hunk 11]
```diff
@@ -49,6 +49,38 @@ export function htmlToParentElement(html) {
...
+        return parser.parseFromString(html, 'text/html').body;
+    }
+    const doc = document.implementation.createHTMLDocument('');
+    if (html.toLowerCase().indexOf('<!doctype') > -1) {
+        // eslint-disable-next-line no-unsanitized/property
+        doc.documentElement.innerHTML = html;
+    } else {
+        // eslint-disable-next-line no-unsanitized/property
+        doc.body.innerHTML = html;
+    }
```
Comment (jnatalzia) [inline src/js/utils/dom.js]: Never seen these objects/methods before! this is some _awesome_ polyfilling. 👏👏

Comment (robwalch) [reply src/js/utils/dom.js]: Yeah. Adapted from the example at https://developer.mozilla.org/en-US/docs/Web/API/DOMParser

**Pull request merged**